Implement transaction negotiation message resending in the wallet

philipr-za opened the pull request:
## Description
Currently the logic for sending transactions and negotiating them between parties is not very robust. Each stage in the negotiation is sent only once. When sending we first try to send the message directly and if that fails, or discovery needs to occur, the message is sent indirectly via Store and Forward (SAF). We hoped to get SAF 100% reliable but due to reasons laid out in https://github.com/tari-project/tari/issues/2137 it is not tenable. 

This PR adds in the logic to periodically resends the initial Transaction Sender Message and the Transaction Reply in order to illicit the next step in the negotiation protocol from the counterparty. The PR also adds the logic to respond to repeated messages but with a cool down period in which repeats will be ignored to protect against DoS attacks. 

The PR also adds in a new comms level message to inform the counterparty if a transaction protocol has been cancelled. Initially this is sent if the Sender cancels an in progress transaction so that the receiver knows to stop expecting it. This message is also sent if a Transaction Reply is received for a cancelled message to let the Received know about the cancellation.

In order to stop a wallet resending these message indefinitely a transaction will be cancelled if it hasn’t resolved after a long timeout period (Default is being set to 3 days).

A large portion of the PR is tests and some code reorganization so its not as big as it looks.

## Motivation and Context
Closes https://github.com/tari-project/tari/issues/2137
Closes https://github.com/tari-project/tari/issues/1804

## How Has This Been Tested?
Tests are provided that test:
1. A sender repeats sending the initial message if no response is received from the receiver
2. A receiver repeats sending the Reply in no response is received from the receiver
3. For both parties that the repeated message is sent on startup if the period has elapsed.
4. For both parties that they can handle receiving the repeated messages and respond appropriately
5. Check that both parties ignore repeated messages sent before the cool down has elapsed
6. Check the cancellation messages are sent by the Sender when a transaction is cancelled
7. Check that receiver acts correctly on receiving a cancellation message
8. Check that the transaction is cancelled by both parties after the timeout is reached during operation or on startup.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
* [ ] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [x] New Tests
* [ ] Documentation

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
* [x] I'm merging against the `development` branch.
* [x] I ran `cargo-fmt --all` before pushing.
* [ ] I have squashed my commits into a single commit.
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.

sdbondi commented on base_layer/p2p/src/proto/tari.p2p.message_type.rs:
```diff
@@ -15,8 +15,9 @@ pub enum TariMessageType {
     MempoolResponse = 72,
-    /// -- DAN Messages --
     TransactionFinalized = 73,
+    /// -- DAN Messages --
+    TransactionCancelled = 74,
```
```suggestion
    TransactionCancelled = 74,
     // -- DAN Messages --
```

sdbondi replied:
Clicked "Add single comment" instead of "Start a review" :P

philipr-za replied:
I keep trying to do that but that is the generated file

sdbondi replied:
Ah I missed that - nevermind then :)

sdbondi commented on base_layer/wallet/src/lib.rs:
```diff
@@ -1,4 +1,4 @@
-#![recursion_limit = "1024"]
```
Oof - I think that this may [be resolved in rust >=1.45 (July 2020)](https://github.com/rust-lang/futures-rs/issues/1917#issuecomment-548135409) - so a compiler upgrade will eventually remove the need to set recursion limit.

philipr-za replied:
Ja, might also be the trigger to split up the Tx Service a bit though need to figure out how to do it without adding more wiring

sdbondi commented on base_layer/wallet/src/transaction_service/protocols/transaction_receive_protocol.rs:
```diff
@@ -406,10 +250,75 @@ where TBackend: TransactionBackend + Clone + 'static
+        };
+
+        if resend {
+            if let Err(e) = send_transaction_reply(
+                inbound_tx.clone(),
+                self.resources.outbound_message_service.clone(),
+                self.resources.config.direct_send_timeout,
+            )
+            .await
+            {
+                warn!(
+                    target: LOG_TARGET,
+                    "Error resending Transaction Reply (TxId: {}): {:?}", self.id, e
+                );
+            }
+            self.resources
+                .db
+                .increment_send_count(self.id)
+                .await
+                .map_err(|e| TransactionServiceProtocolError::new(self.id, TransactionServiceError::from(e)))?;
```
Shouldn't `increment_send_count` only be called if the `send_transaction_reply` succeeds?

philipr-za replied:
I considered it as an attempt was made to send and we will only try again (succeed or fail) again in the next period. Wanted to be careful not to get stuck attempting too many sends as it would thrash a mobile's battery.

sdbondi commented on base_layer/wallet/src/transaction_service/protocols/transaction_receive_protocol.rs:
```diff
@@ -430,6 +339,29 @@ where TBackend: TransactionBackend + Clone + 'static
                         ));
+                    },
+                    () = resend_timeout => {
+                        if let Err(e) = send_transaction_reply(
+                            inbound_tx.clone(),
+                            self.resources.outbound_message_service.clone(),
+                            self.resources.config.direct_send_timeout,
+                        )
+                        .await
+                        {
+                            warn!(
+                                target: LOG_TARGET,
+                                "Error resending Transaction Reply (TxId: {}): {:?}", self.id, e
+                            );
+                        } else {
+                            self.resources
+                                .db
+                                .increment_send_count(self.id)
+                                .await
+                                .map_err(|e| TransactionServiceProtocolError::new(self.id, TransactionServiceError::from(e)))?;
+                        }
```
`match` is more idiomatic

```suggestion
                        match send_transaction_reply(
                            inbound_tx.clone(),
                            self.resources.outbound_message_service.clone(),
                            self.resources.config.direct_send_timeout,
                        )
                        .await
                        {
                        Ok(_) => {
                           self.resources
                                .db
                                .increment_send_count(self.id)
                                .await
                                .map_err(|e| TransactionServiceProtocolError::new(self.id, TransactionServiceError::from(e)))?;
                        },
                        Err(err) => {
                            warn!(
                                target: LOG_TARGET,
                                "Error resending Transaction Reply (TxId: {}): {:?}", self.id, e
                            );
                        
                        }
```

philipr-za replied:
Damnit Stanley, why do you hate the classic C syntax so much?

philipr-za replied:
But ok will update but honestly its because the classic syntax is easier to format manually inside a select! macro

sdbondi replied:
Yeess .. submit to the will of rust

sdbondi reviewed (dismissed):
LGTM - will run a few tests on the branch before approving

sdbondi reviewed (dismissed):
All my stuck transactions on the base node went through after one retry 🎉

philipr-za committed:
Implement transaction negotiation message resending in the wallet

Currently the logic for sending transactions and negotiating them between parties is not very robust. Each stage in the negotiation is sent only once. When sending we first try to send the message directly and if that fails, or discovery needs to occur, the message is sent indirectly via Store and Forward (SAF). We hoped to get SAF 100% reliable but due to reasons laid out in https://github.com/tari-project/tari/issues/2137 it is not tenable. 

This PR adds in the logic to periodically resends the initial Transaction Sender Message and the Transaction Reply in order to illicit the next step in the negotiation protocol from the counterparty. The PR also adds the logic to respond to repeated messages but with a cool down period in which repeats will be ignored to protect against DoS attacks. 

The PR also adds in a new comms level message to inform the counterparty if a transaction protocol has been cancelled. Initially this is sent if the Sender cancels an in progress transaction so that the receiver knows to stop expecting it. This message is also sent if a Transaction Reply is received for a cancelled message to let the Received know about the cancellation.

In order to stop a wallet resending these message indefinitely a transaction will be cancelled if it hasn’t resolved after a long timeout period (Default is being set to 3 days).

A large portion of the PR are tests so its not as big as it looks.
